metashare/api/serializers.py

```diff
@@ -86,13 +86,58 @@ def get_repo_url(self, obj) -> Optional[str]:
         return f"https://github.com/{obj.repo_owner}/{obj.repo_name}"
 
 
+class HookSerializerMixin:
+    def get_matching_repository(self):
+        repo_id = self.validated_data["repository"]["id"]
+        return Repository.objects.filter(repo_id=repo_id).first()
+
+
+class HookRepositorySerializer(HookSerializerMixin, serializers.Serializer):
+    id = serializers.IntegerField()
+
+
 class AuthorCommitSerializer(serializers.Serializer):
     name = serializers.CharField(required=False)
     email = serializers.CharField(required=False)
     username = serializers.CharField(required=False)
     avatar_url = serializers.CharField(required=False)
 
 
+class PrSerializer(serializers.Serializer):
+    merged = serializers.BooleanField()
+    # All other fields are ignored by default.
+
+
+class PrHookSerializer(HookSerializerMixin, serializers.Serializer):
+    action = serializers.CharField()
+    number = serializers.IntegerField()
+    pull_request = PrSerializer()
+    repository = HookRepositorySerializer()
+    # All other fields are ignored by default.
+
+    def _is_merged(self):
+        return (
+            self.validated_data["action"] == "closed"
+            and self.validated_data["pull_request"]["merged"]
+        )
+
+    def _get_matching_task(self, repository):
+        return Task.objects.filter(
+            project__repository=repository, pr_number=self.validated_data["number"]
+        ).first()
+
+    def process_hook(self):
+        repository = self.get_matching_repository()
+        if not repository:
+            raise NotFound("No matching repository.")
+
+        if self._is_merged():
+            task = self._get_matching_task(repository)
+            if task is None:
+                return
+            task.finalize_status_completed()
```

Comment:
I think we also want to set `task.has_unmerged_commits = False`, whether here or in the `task.finalize_status_completed()` method.